Refactor Conda Recipes

milljm opened the pull request:
Refactor the recipes to be more loose with dependencies in some areas,
and more strict in others. Namely the compilers.

moose-tools no longer depends on any other moose- package. It is our
hope that cutting ties with moose-mpich will help when users install
other packages.

moose-mpich has a _very_ strict dependency set (compilers, hdf5, mesa).

[edit]:
moose-libmesh-vtk; remove the Qt/PyQt dependency. This will sever ties
with many dependencies that was hindering additional package installs.
This also affords us a single moose-libmesh-vtk build.

moose- petsc, libmesh-vtk, and libmesh no longer depend on a hash_build_#
These packages will now pull whatever is the latest version and available
at the time. After contemplating how our build system works, this should
just always get the right package, locally built, or /public. I believe
it may also be possible to remove the version entirely (do that later).

Add moose-gperftools [edit]: renamed to moose-pprof
Add moose-autojump

In order for all this to work, we had to break away from a common
symlink to conda_build_config.yaml. There remains a few places where we
can still have a common config, but I opted to move forward with
separating everything.

One caveat, is the necessity to start over. No one will be able to
upgrade to any of these packages which depend on moose-mpich. Something
about Python/VTK/PyQt's deps. Once a person has the latest moose-mpich
package, things will start to fall into place.

Refs: #21247 #21111

Commit:
Refactor Conda Recipes

Refactor the recipes to be more loose with dependencies in some areas,
and more strict in others. Namely the compilers.

moose-tools no longer depends on any other moose- package. It is our
hope that cutting ties with moose-mpich will help when users install
other packages.

moose-mpich has a _very_ strict dependency set (compilers, hdf5, mesa).
This should help keep libmesh-vtk in line with the pesky PyQt dependency
set.

moose- petsc, libmesh-vtk, and libmesh no longer depend on a hash_build_#
These packages will now pull whatever is the latest version and available
at the time. After contemplating how our build system works, this should
just always get the right package, locally built, or /public. I believe
it may also be possible to remove the version entirely (do that later).

Add moose-gperftools
Add moose-autojump

In order for all this to work, we had to break away from a common
symlink to conda_build_config.yaml. There remains a few places where we
can still have a common config, but I opted to move forward with
separating everything.

One caveat, is the necessity to start over. No one will be able to
upgrade to any of these packages which depend on moose-mpich. Something
about Python/VTK/PyQt's deps. Once a person has the latest moose-mpich
package, things will start to fall into place.

Refs: #21247 #21111

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1081756/) on d35042d wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1081756/) on d35042d wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1081774/) on d35042d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21324/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1081783/) on d35042d wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ad2d41</th>
      <th colspan=4>#21324 d35042</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.46%</td>
      <td>83.46%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>77427</td>
      <td>77431</td>
      <td>+4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15346</td>
      <td>15342</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21324/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21324/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21324/coverage/xfem)

</details>



This comment will be updated on new commits.

milljm commented on conda/libmesh-vtk/build.sh:
```diff
@@ -1,22 +1,15 @@
   QT="NO"
 fi
-
+echo "CC, CXX $CC $CXX"
```
remove this

cticenhour commented on conda/gperftools/LICENSE.txt:
```diff
@@ -0,0 +1,13 @@
+BSD 3-clause license
+Copyright (c) 2015-2018, conda-forge
+All rights reserved.
```
So we have to include a `conda-forge` license with a Google open source product? Isn't there a Google license we need to use?

milljm replied:
hmm yeah probably. I'll be sure to figure that out

cticenhour replied:
So we're using their recipe you mean, and then overwriting? In that case I could see providing both open source licenses in a single file.

cticenhour commented on conda/gperftools/build.sh:
```diff
@@ -0,0 +1,34 @@
+# Get an updated config.sub and config.guess
+cp $BUILD_PREFIX/share/gnuconfig/config.* .
+if [ `uname` == "Darwin" ]; then
+    TEMP_CXXFLAGS=${CXXFLAGS//-std=c++[0-9][0-9]}
+    # gperftools uses random_shuffle (removed in c++17)
+    ACTIVATION_CXXFLAGS=${TEMP_CXXFLAGS%%-fdebug-prefix-map*}-std=c++14
```
Kinda glad this little shuffle is useful elsewhere 😂

cticenhour commented on conda/gperftools/info/has_prefix:
```diff
@@ -0,0 +1 @@
+/opt/anaconda1anaconda2anaconda3 text etc/activate.d/activate_moose-gperftools.sh
```
Can you explain what's going on here?

milljm replied:
This is a post process operation that alters any paths it sees in the specified file path (`etc/activate.d/activate_moose-gperftools.sh`), having a prefix with all those:
```pre
_placehold_placehold_placehold_placehold*
```
paths, changed to the correct final path of $CONDA_PREFIX. This is so when all is said and done, the end user john_doe who installed Conda to some location, has a working activation script:

```bash
/Users/john_doe/libs/conda3/envs/moose/etc/conda/activate.d/activate_moose-gperftools.sh
export GPERF_DIR=/Users/john_doe/libs/conda3/envs/moose
```
Since we don't know what that will be before hand, this is basically how Conda does it :). I probably should just have linked the following. Will explain better than I can:
Ref: https://docs.conda.io/projects/conda-build/en/latest/resources/make-relocatable.html

cticenhour replied:
Looks good - thanks for the explanation!

cticenhour commented on conda/mpich/conda_build_config.yaml:
```diff
@@ -1,245 +1,63 @@
-## changes made will require additional changes to items listed above
-## your change (except where otherwise noted)
-
-# NCRC dependency
-moose_tools:
-  - moose-tools 2022.04.18
-
-# NCRC default Python
-moose_ncrc_python:
-  - python 3.9
```
Removing this won't affect anything? I don't know anything about NCRC needs.

cticenhour commented on conda/petsc/run_test.sh:
```diff
@@ -1,6 +0,0 @@
-#!/bin/bash
-set -e
-
-export PETSC_DIR=${PREFIX}
-
-pkg-config --cflags PETSc | grep -v isystem
```
Just double-checking that this is no longer needed. I don't see any changes in the `meta.yaml` regarding testing, so I am assuming we should have removed this a while back?

milljm replied:
Oh, yeah we can probably put this back. :/ not sure how/why I deleted it.

cticenhour reviewed (commented):
The GitHub UI shows that `conda/pyhit/build.sh` is an empty file that somehow changes size. When I click on "View File", it takes me to your refactor commit (or at least a version of it) that has content but claims to not be on any branches.....can you double-check that this file exists on your local branch?

Commit:
Remove PyQt from libmesh-vtk

Drop PyQt from libmesh-vtk requirements to ease up on the dependency
restrictions. PyQt is simply not a good thing to have a dependency for.

We will need (if we choose) to create a moose-peacock package to allow
for Peacock functionality.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1083827/) on 70be059 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1083827/) on 70be059 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1083827/) on 70be059 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
Addressing PR concerns

Rename gpertoolfs package to accommodate Apache license for pprof.

Bring back petsc run_tests.sh.

Fix test path for libmesh-vtk. Odd that lib64 is no longer there? Must
be an additional package that triggers that path?

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1084195/) on 8d10fad wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1084195/) on 8d10fad wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
need graphviz for pprof png writes

add graphviz to moose-pprof's runtime dependencies.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1084727/) on 63f8d82 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1084727/) on 63f8d82 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1084727/) on 63f8d82 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1084727/) on 63f8d82 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1084727/) on 63f8d82 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

cticenhour commented on conda/pprof/LICENSE:
```diff
@@ -0,0 +1,13 @@
+BSD 3-clause license
+Copyright (c) 2015-2018, conda-forge
```
OK - I like the move to pprof here....So we've moved to using conda forge gperftools and then downloading and building pprof? Shouldn't this then be the google license?

cticenhour commented on conda/libmesh-vtk/build.sh:
```diff
@@ -27,41 +12,24 @@ cmake .. -G "Ninja" \
     -DCMAKE_PREFIX_PATH:PATH=${VTK_PREFIX} \
     -DCMAKE_INSTALL_PREFIX:PATH=${VTK_PREFIX} \
     -DCMAKE_INSTALL_RPATH:PATH=${VTK_PREFIX}/lib \
-    -DCMAKE_INSTALL_LIBDIR=lib \
```
In your commit message changing the libmesh-vtk package tests in meta.yaml, you wondered why lib needed to be changed to lib64. I think this is why. Why did you remove this?

Commit:
Addressing comments; fix pprof license

Add Apache liscense to pprof recipe.
Specify lib dir is `lib`.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1085317/) on ed14fc2 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1085317/) on ed14fc2 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1085317/) on ed14fc2 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
Fix read-only go installed files

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1085426/) on c28ea25 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1085426/) on c28ea25 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1085426/) on c28ea25 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
Do the modcache cleaning properly...

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1085524/) on 73036e2 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1085524/) on 73036e2 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops
